### ISSUE TITLE
Resolve Pitest Issues - JavadocMethodCheck (1)

### ISSUE THREAD

**Issue by rnveach** [checkstyle/checkstyle#7922]
Child issue of https://github.com/checkstyle/checkstyle/issues/7797 ,

https://github.com/checkstyle/checkstyle/blob/f9131f9536c71d963c83aab1550b7a0371aa9ff5/.ci/pitest.sh#L182

**Comment by sulkykookie** [comment]
I'm on it.

**Comment by sulkykookie** [comment]
### 1. Pitest Report
https://sulkykookie.github.io/issue4/202003261154/com.puppycrawl.tools.checkstyle.checks.javadoc/JavadocMethodCheck.java.html
Surviving Mutation: ```negated conditional → SURVIVED``` at line 1042

### 2. Mutated Branch
Hardcoded commit: a155943
Since the surviving mutation is ```negated conditional → SURVIVED```, it would be better to get rid of the ```if``` statement completely.

### 3. Regression Report
https://sulkykookie.github.io/issue4/diff/
_Zero differences_

### 4. Code Analysis
Since the survived mutation says that even if the conditional is negated, tests work fine, it is better to get rid of the ```if``` statement completely.
It also makes sense as if we look into the ```processTree``` method, ```if``` statement is checking whether there is an outer class or not. Let us assume there is no outer class, then the inner class would itself be the outer class.
So in order for the inner class to exist, there must be an existing outer class as well, and therefore it is not necessary to check if the filename is not empty as then the filenames would end up being ```$inner.class``` instead of ```outer$inner.class```.

**Comment by sulkykookie** [comment]
@strkkk Please review

**Comment by rnveach** [comment]
Pitest and branch look good.

> https://sulkykookie.github.io/issue4/diff/android-launcher/index.html

Regression only has the default module and this check has many properties. https://checkstyle.org/config_javadoc.html#JavadocMethod_Properties

We expand regression to more different instances as described in https://github.com/checkstyle/contribution/tree/master/checkstyle-tester#checkstyle-pitest-regression .

**Comment by sulkykookie** [comment]
@rnveach check now

**Comment by rnveach** [comment]
Multiple config instances can be run inside 1 run. You don't need to split it out for each config.

> Since the survived mutation says that even if the conditional is negated, tests work fine, it is better to get rid of the if statement completely.

I assume you mean to remove the if and leave the code inside it? Why does removing the condition but leaving the code have no effect? We don't care if class name has a `$` or not?

> So in order for the inner class to exist, there must be an existing outer class as well, and therefore it is not necessary to check if the filename is not empty.
> There is no use in prepending a $ to inner class' name when there isn't an outer class.

It is checking class name, not file name. So are you saying we don't even have an outer class? Are we tracking something wrong.


Please provide links to where you are talking about in the code so I can follow along. I am not deeply familar with all parts of this check. It is not easy following your logic you described. I want to be proven to that these is code is absolutely not needed. It may be beneficial to take me through a case, line by line.

**Comment by sulkykookie** [comment]
> Multiple config instances can be run inside 1 run. You don't need to split it out for each config.
> 
> > Since the survived mutation says that even if the conditional is negated, tests work fine, it is better to get rid of the if statement completely.
> 
> I assume you mean to remove the if and leave the code inside it? Why does removing the condition but leaving the code have no effect? We don't care if class name has a `$` or not?
> 
> > So in order for the inner class to exist, there must be an existing outer class as well, and therefore it is not necessary to check if the filename is not empty.
> > There is no use in prepending a $ to inner class' name when there isn't an outer class.
> 
> It is checking class name, not file name. So are you saying we don't even have an outer class? Are we tracking something wrong.
> 
> Please provide links to where you are talking about in the code so I can follow along. I am not deeply familar with all parts of this check. It is not easy following your logic you described. I want to be proven to that these is code is absolutely not needed. It may be beneficial to take me through a case, line by line.

I'm gonna look more into the code.

About the regression report, as there are so many properties with so many possible values, there can be so many permutations. How do I proceed about that?

**Comment by rnveach** [comment]
> About the regression report, as there are so many properties with so many possible values, there can be so many permutations. How do I proceed about that?

Since you are going to examine the code more, I would use the knowledge you gain to narrow down which properties matter and which don't. Your first batch was a pretty good start and shows blindly doing more regression may not get you anymore answers than you have know. Regression will not be the answer to all pitest problems. It could be a case where we need specific set of properties or a input file that is pretty uncommon. I have seen pitest cases where a package name had to be the same as a class name to kill the mutation.

**Comment by sulkykookie** [comment]
> > About the regression report, as there are so many properties with so many possible values, there can be so many permutations. How do I proceed about that?
> 
> Since you are going to examine the code more, I would use the knowledge you gain to narrow down which properties matter and which don't. Your first batch was a pretty good start and shows blindly doing more regression may not get you anymore answers than you have know. Regression will not be the answer to all pitest problems. It could be a case where we need specific set of properties or a input file that is pretty uncommon. I have seen pitest cases where a package name had to be the same as a class name to kill the mutation. I can't imagine you would find anything like that in real projects.

I have a small doubt regarding the traversal of the tree.
So suppose the case:
```
class Outer {
    class Inner {
    }
}
```

In this case, the ```beginTree(DetailAST ast)``` function starts and
https://github.com/checkstyle/checkstyle/blob/fab949ad62a4ec520afe4924c181a1c304533d4c/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java#L382
executes making ```currentClassName``` as nothing. So when the ast of the Outer class begins, 
https://github.com/checkstyle/checkstyle/blob/fab949ad62a4ec520afe4924c181a1c304533d4c/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java#L391
is executed as its a CLASS_DEF and which in turn makes the ```currentClassName``` "Outer", but since there is a node of ```Inner``` class, the ```visitToken``` method executes for it as well and it makes the ```currentClassName``` "Outer$Inner".
Now the ```leaveToken``` method of Inner class executed and the ```currentClassName``` becomes "Outer" according to
https://github.com/checkstyle/checkstyle/blob/fab949ad62a4ec520afe4924c181a1c304533d4c/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java#L407

Then afterwards ```leaveToken``` executes for Outer class and since there is no "." or "$" in the ```currentClassName``` variable name, it makes ```currentClassName``` "" (blank). How is that possible?

**Comment by sulkykookie** [comment]
@rnveach please check

**Comment by sulkykookie** [comment]
@rnveach I can't proceed without knowing where im lacking please help me with this.

**Comment by sulkykookie** [comment]
@rnveach 

From what I have gathered from the code until now is the only property requiring the ```currentClassName``` variable is ```validateThrows```. As can be seen in the code:
https://github.com/checkstyle/checkstyle/blob/834b505b9938e93f6950dd20a13e39e295e5ac86/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java#L485

which further calls ```getThrows``` and ```getThrowed``` method both requiring the ```currentClassName``` variable.
However, once we get the ```classInfo``` by this ```currentClassName``` variable, it calls ```checkThrowsTags``` method:
https://github.com/checkstyle/checkstyle/blob/834b505b9938e93f6950dd20a13e39e295e5ac86/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java#L486

which further needs ```currentClassName``` in:
https://github.com/checkstyle/checkstyle/blob/834b505b9938e93f6950dd20a13e39e295e5ac86/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java#L925

which just checks whether the class names are same or not:
https://github.com/checkstyle/checkstyle/blob/834b505b9938e93f6950dd20a13e39e295e5ac86/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java#L953

**Bottom Line:** It shouldn't matter what the class name is, at the moment, that is, even the outer class' name can have a $ before its name in the ```currentClassName``` variable and it should not matter as it's just going to check the equality of the class names as shown above that is why that ```if``` can be removed from the check's java files, keeping its body only.

**Comment by romani** [comment]
fix is meged.
